File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """

  defguard is_even(input) when is_integer(input) and rem(input, 2) == 0
  defguard is_odd(input) when is_integer(input) and rem(input, 2) != 0

  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(input) when is_integer(input) and input > 0, do: calc(input, 0)

  defp calc(1, step), do: step
  defp calc(input, step) when is_even(input), do: calc(div(input, 2), step + 1)
  defp calc(input, step) when is_odd(input), do: calc((input * 3 + 1), step + 1)
end

File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let steps (number: int): int option = 
    let rec tailSteps (number: int) (acc: int) =
        if number = 1 then Some acc
        elif number % 2 = 0 then (tailSteps (number/2) (acc+1))
        else (tailSteps (3*number + 1) (acc+1))
    if number <= 0 then None
    else tailSteps number 0